crates/test/macros/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+                        is_test = true;
+                        false
+                    } else {
+                        true
+                    }
+                });
+
+                if is_test {
+                    if !f.sig.inputs.is_empty() {
+                        return Err(parse::Error::new(
+                            f.sig.ident.span(),
+                            "`#[test]` function must have signature `fn()`",
+                        ));
+                    }
+
+                    tests.push(f);
+                } else {
+                    untouched_tokens.push(Item::Fn(f));
+                }
+            }
+            _ => {
+                untouched_tokens.push(item);
+            }
+        }
+    }
+
+    let ident = module.ident;
+    let test_count = tests.len();
+    let test_names = tests.iter().map(|test| &test.sig.ident);
+
+    Ok(quote!(
+        #[cfg(test)]
+        pub(crate) mod #ident {
+            #(#untouched_tokens)*
+
+            #(#tests)*
+
+            pub(crate) fn __test_list() -> impl Iterator<Item = fn()> {
+                const TEST_LIST: [fn(); #test_count] = [#(#test_names), *];
+                TEST_LIST.into_iter()
+            }
+        }
+    )
+    .into())
+}
+
+#[proc_macro]
+pub fn tests_runner(args: TokenStream) -> TokenStream {
+    match tests_runner_impl(args) {
+        Ok(ts) => ts,
+        Err(e) => e.to_compile_error().into(),
+    }
+}
+
+fn tests_runner_impl(args: TokenStream) -> parse::Result<TokenStream> {
+    let args: syn::ExprArray = syn::parse(args)?;
+
+    let test_suites = args
+        .elems
+        .into_iter()
+        .map(|attr| quote!(let ret = ret.chain(#attr::__test_list());))
+        .collect::<Vec<_>>();
+
+    Ok(quote!(
+        #[cfg(test)]
+        mod __test_runner {
+            // Required for panic handler
+            extern crate flipperzero_rt;
+
+            // Required for allocator
+            #[cfg(feature = "alloc")]
+            extern crate flipperzero_alloc;
+
+            use core::time::Duration;
+
+            use flipperzero_rt::{entry, manifest};
+            use flipperzero_sys as sys;
+
+            use crate::{furi::thread, println};
+
+            manifest!(name = "Rust Unit Tests");
+            entry!(main);
+
+            fn test_list() -> impl Iterator<Item = fn()> {
+                let ret = ::core::iter::empty();
+                #(#test_suites)*
+                ret
+            }
+
+            // Test runner entry point
+            fn main(_args: *mut u8) -> i32 {
+                let heap_before = unsafe { sys::memmgr_get_free_heap() };
+                let cycle_counter = unsafe { sys::furi_get_tick() };
+
+                let failed = 0;
+                for test in test_list() {
+                    // When the test assertions fail, we can't catch the panics.
+                    // TODO: Replace panics with Results.
+                    test();
```

Comment:
The current state of this PR does successfully run tests on the Flipper Zero. However, if any test assertion fails, the unit test FAP crashes so we won't get any results. Given how `flipperzero_rt` intercepts the panic handler, and `catch_unwind` isn't available anyway without `std`, I think the best solution is to require tests to return `Result<(), &'static str>`, and then maybe use the proc macro to rewrite the assertion-based tests if possible.

Reply:
Intercepting the `assert` macros works very nicely!